mobile/src/main/java/org/openhab/habdroid/ui/OpenHABWidgetAdapter.java

```diff
@@ -55,921 +51,1075 @@
+                mRefreshRate = 0;
+            } else {
+                // Widget URL may be relative, add base URL if that's the case
+                Uri uri = Uri.parse(widget.getUrl());
+                if (uri.getScheme() == null) {
+                    uri = Uri.parse(mConnectionInfo.baseUrl + widget.getUrl());
                 }
-                break;
-            default:
-                if (labelTextView != null)
-                    labelTextView.setText(openHABWidget.getLabel());
-                break;
+                mImageView.setImageUrl(uri.toString(), mConnectionInfo.userName,
+                        mConnectionInfo.password, false);
+                mRefreshRate = widget.getRefresh();
+            }
         }
-        LinearLayout dividerLayout = (LinearLayout) widgetView.findViewById(R.id.listdivider);
-        if (dividerLayout != null) {
-            if (position < this.getCount() - 1) {
-                if (this.getItemViewType(position + 1) == TYPE_FRAME) {
-                    dividerLayout.setVisibility(View.GONE); // hide dividers before frame widgets
-                } else {
-                    dividerLayout.setVisibility(View.VISIBLE); // show dividers for all others
-                }
-            } else { // last widget in the list, hide divider
-                dividerLayout.setVisibility(View.GONE);
+
+        @Override
+        public void start() {
+            if (mRefreshRate > 0) {
+                mImageView.setRefreshRate(mRefreshRate);
+            } else {
+                mImageView.cancelRefresh();
             }
         }
-        return widgetView;
-    }
 
-    @Override
-    public int getViewTypeCount() {
-        return TYPES_COUNT;
+        @Override
+        public void stop() {
+            mImageView.cancelRefresh();
+        }
     }
 
-    @Override
-    public int getItemViewType(int position) {
-        OpenHABWidget openHABWidget = getItem(position);
-        if (openHABWidget.getType().equals("Frame")) {
-            return TYPE_FRAME;
-        } else if (openHABWidget.getType().equals("Group")) {
-            return TYPE_GROUP;
-        } else if (openHABWidget.getType().equals("Switch")) {
-            if (openHABWidget.hasMappings()) {
-                return TYPE_SECTIONSWITCH;
-            } else if (openHABWidget.getItem() != null) {
-                if (openHABWidget.getItem().getType() != null) {
-                    //RollerShutterItem changed to RollerShutter in later builds of OH2
-                    if ("RollershutterItem".equals(openHABWidget.getItem().getType()) ||
-                            "Rollershutter".equals(openHABWidget.getItem().getType()) ||
-                            "Rollershutter".equals(openHABWidget.getItem().getGroupType()))
-                        return TYPE_ROLLERSHUTTER;
-                    else
-                        return TYPE_SWITCH;
-                } else
-                    return TYPE_SWITCH;
+
```

Comment:
Is there a reason why you have chosen OnTouchListener over OnClickListener? From what I read (https://stackoverflow.com/a/12811038/9053008) onTouch does not work with mouse and keyboard.

Reply:
Who is that 'you'? ;-) I just took what was already there, the OnTouchListener was part of that. FWIW, I wondered about the very same question myself though. In some cases (e.g. switch) there are even more specific change listeners we could use.

Reply:
Oh, I haven't checked what was before your change...